File: grey-haven-plugins/core/skills/llm-project-development/templates/pipeline-template.py
"""
Grey Haven LLM Pipeline Template - Python

Copy this template and customize for your use case.

Usage:
1. Copy to app/pipelines/your_pipeline.py
2. Define your input/output models
3. Implement the abstract methods
4. Create API route or service integration
"""

from abc import ABC, abstractmethod
from pathlib import Path
from datetime import datetime
from typing import TypeVar, Generic
from uuid import UUID
import re
import json

from pydantic import BaseModel, Field
from sqlalchemy import select, update
from sqlalchemy.ext.asyncio import AsyncSession
from anthropic import Anthropic

# ============================================================================
# 1. DEFINE YOUR OUTPUT MODEL
# ============================================================================

class YourOutput(BaseModel):
    """Parsed output from LLM."""
    summary: str
    category: str = Field(pattern="^(type_a|type_b|type_c)$")
    confidence: float = Field(ge=0.0, le=1.0)
    # Add more fields as needed


# ============================================================================
# 2. DEFINE YOUR INPUT MODEL (or use SQLModel)
# ============================================================================

class YourInput(BaseModel):
    """Input data structure."""
    id: UUID
    tenant_id: UUID
    title: str
    content: str
    # Add your input fields


# ============================================================================
# 3. BASE PIPELINE CLASS (Don't modify)
# ============================================================================

TInput = TypeVar("TInput")
TOutput = TypeVar("TOutput")


class BasePipeline(ABC, Generic[TInput, TOutput]):
    """Base class for LLM pipelines with file-based caching."""

    def __init__(
        self,
        tenant_id: str,
        cache_dir: Path | None = None,
        model: str = "claude-sonnet-4-20250514"
    ):
        self.tenant_id = tenant_id
        self.model = model
        self.client = Anthropic()
        self.cache_dir = cache_dir or Path(".cache") / "pipelines" / tenant_id

    @abstractmethod
    async def acquire(self, item_id: str) -> TInput:
        """Stage 1: Get input data from database."""
        pass

    @abstractmethod
    def prepare(self, input_data: TInput) -> str:
        """Stage 2: Format prompt for LLM."""
        pass

    @abstractmethod
    def parse(self, response: str) -> TOutput:
        """Stage 4: Parse LLM response."""
        pass

    @abstractmethod
    async def render(self, item_id: str, output: TOutput) -> TOutput:
        """Stage 5: Save output to database."""
        pass

    async def process(self, prompt: str, cache_key: str) -> str:
        """Stage 3: Call LLM with file-based caching."""
        cache_path = self.cache_dir / f"{cache_key}.json"

        # Check cache
        if cache_path.exists():
            cached = json.loads(cache_path.read_text())
            print(f"[cache hit] {cache_key}")
            return cached["response"]

        # Call LLM
        print(f"[llm call] {cache_key} using {self.model}")
        response = self.client.messages.create(
            model=self.model,
            max_tokens=1024,
            messages=[{"role": "user", "content": prompt}]
        )

        text = response.content[0].text

        # Cache response
        self.cache_dir.mkdir(parents=True, exist_ok=True)
        cache_path.write_text(json.dumps({
            "response": text,
            "model": self.model,
            "input_tokens": response.usage.input_tokens,
            "output_tokens": response.usage.output_tokens,
            "timestamp": datetime.utcnow().isoformat(),
        }))

        return text

    async def run(self, item_id: str) -> TOutput:
        """Execute full pipeline."""
        input_data = await self.acquire(item_id)
        prompt = self.prepare(input_data)
        response = await self.process(prompt, item_id)
        output = self.parse(response)
        return await self.render(item_id, output)

    async def run_batch(self, item_ids: list[str]) -> dict[str, TOutput]:
        """Run pipeline for multiple items."""
        results = {}
        for item_id in item_ids:
            try:
                output = await self.run(item_id)
                results[item_id] = output
            except Exception as e:
                print(f"[error] {item_id}: {e}")
        return results


# ============================================================================
# 4. YOUR PIPELINE IMPLEMENTATION
# ============================================================================

# TODO: Import your SQLModel
# from app.models.your_model import YourModel


class YourPipeline(BasePipeline[YourInput, YourOutput]):
    """Your custom pipeline implementation."""

    def __init__(self, tenant_id: str, db: AsyncSession, **kwargs):
        super().__init__(tenant_id, **kwargs)
        self.db = db

    async def acquire(self, item_id: str) -> YourInput:
        """Stage 1: Get input data from database."""
        # TODO: Replace with your model and table
        result = await self.db.execute(
            select(YourModel).where(
                YourModel.tenant_id == self.tenant_id,
                YourModel.id == item_id
            )
        )
        item = result.scalar_one_or_none()

        if not item:
            raise ValueError(
                f"Item {item_id} not found for tenant {self.tenant_id}"
            )

        return item

    def prepare(self, input_data: YourInput) -> str:
        """Stage 2: Format prompt for LLM."""
        # TODO: Customize your prompt
        return f"""Analyze this content.

TITLE: {input_data.title}

CONTENT:
{input_data.content}

I will parse this programmatically. Respond with valid JSON:
{{
  "summary": "Brief summary (1-2 sentences)",
  "category": "type_a" | "type_b" | "type_c",
  "confidence": 0.0 to 1.0
}}

Ensure JSON is complete and parseable."""

    def parse(self, response: str) -> YourOutput:
        """Stage 4: Parse LLM response."""
        match = re.search(r'\{[\s\S]*\}', response)
        if not match:
            raise ValueError("No JSON found in response")

        try:
            return YourOutput.model_validate_json(match.group())
        except Exception as e:
            raise ValueError(f"Failed to parse response: {e}")

    async def render(self, item_id: str, output: YourOutput) -> YourOutput:
        """Stage 5: Save output to database."""
        # TODO: Replace with your update query
        await self.db.execute(
            update(YourModel)
            .where(
                YourModel.tenant_id == self.tenant_id,
                YourModel.id == item_id
            )
            .values(
                summary=output.summary,
                category=output.category,
                processed_at=datetime.utcnow(),
                updated_at=datetime.utcnow()
            )
        )
        await self.db.commit()

        return output


# ============================================================================
# 5. FACTORY FUNCTION
# ============================================================================

def create_your_pipeline(
    tenant_id: str,
    db: AsyncSession,
    model: str | None = None
) -> YourPipeline:
    """Create pipeline instance."""
    kwargs = {}
    if model:
        kwargs["model"] = model
    return YourPipeline(tenant_id, db, **kwargs)


# ============================================================================
# 6. USAGE EXAMPLE
# ============================================================================

"""
# In API route:
from app.pipelines.your_pipeline import create_your_pipeline

@router.post("/items/{item_id}/process")
async def process_item(
    item_id: UUID,
    db: AsyncSession = Depends(get_db),
    tenant: Tenant = Depends(get_current_tenant),
):
    pipeline = create_your_pipeline(str(tenant.id), db)
    result = await pipeline.run(str(item_id))
    return {"success": True, "result": result.model_dump()}


# Batch processing:
@router.post("/items/process-batch")
async def process_items_batch(
    item_ids: list[UUID],
    background_tasks: BackgroundTasks,
    db: AsyncSession = Depends(get_db),
    tenant: Tenant = Depends(get_current_tenant),
):
    async def process():
        pipeline = create_your_pipeline(str(tenant.id), db)
        await pipeline.run_batch([str(id) for id in item_ids])

    background_tasks.add_task(process)
    return {"success": True, "queued": len(item_ids)}
"""

File: grey-haven-plugins/core/skills/llm-project-development/templates/pipeline-template.ts
/**
 * Grey Haven LLM Pipeline Template - TypeScript
 *
 * Copy this template and customize for your use case.
 *
 * Usage:
 * 1. Copy to lib/pipelines/your-pipeline.ts
 * 2. Define your input/output types
 * 3. Implement the abstract methods
 * 4. Create API route or service integration
 */

import { existsSync, mkdirSync, writeFileSync, readFileSync } from "fs";
import { join } from "path";
import Anthropic from "@anthropic-ai/sdk";
import { z } from "zod";
import { db } from "@/db";
import { eq, and } from "drizzle-orm";

// ============================================================================
// 1. DEFINE YOUR OUTPUT SCHEMA
// ============================================================================

const YourOutputSchema = z.object({
  // Define your expected output structure
  summary: z.string(),
  category: z.enum(["type_a", "type_b", "type_c"]),
  confidence: z.number().min(0).max(1),
  // Add more fields as needed
});

type YourOutput = z.infer<typeof YourOutputSchema>;

// ============================================================================
// 2. DEFINE YOUR INPUT TYPE
// ============================================================================

interface YourInput {
  id: string;
  tenant_id: string;
  // Add your input fields
  title: string;
  content: string;
}

// ============================================================================
// 3. BASE PIPELINE CLASS (Don't modify)
// ============================================================================

abstract class BasePipeline<TInput, TOutput> {
  protected client: Anthropic;
  protected cacheDir: string;
  protected tenantId: string;
  protected model: string;

  constructor(
    tenantId: string,
    options: { cacheDir?: string; model?: string } = {}
  ) {
    this.tenantId = tenantId;
    this.client = new Anthropic();
    this.model = options.model || "claude-sonnet-4-20250514";
    this.cacheDir =
      options.cacheDir ||
      join(process.cwd(), ".cache", "pipelines", tenantId);
  }

  abstract acquire(id: string): Promise<TInput>;
  abstract prepare(input: TInput): string;
  abstract parse(response: string): TOutput;
  abstract render(id: string, output: TOutput): Promise<TOutput>;

  protected async process(prompt: string, cacheKey: string): Promise<string> {
    const cachePath = join(this.cacheDir, `${cacheKey}.json`);

    if (existsSync(cachePath)) {
      console.log(`[cache hit] ${cacheKey}`);
      return JSON.parse(readFileSync(cachePath, "utf-8")).response;
    }

    console.log(`[llm call] ${cacheKey}`);
    const response = await this.client.messages.create({
      model: this.model,
      max_tokens: 1024,
      messages: [{ role: "user", content: prompt }],
    });

    const text =
      response.content[0].type === "text" ? response.content[0].text : "";

    mkdirSync(this.cacheDir, { recursive: true });
    writeFileSync(
      cachePath,
      JSON.stringify({
        response: text,
        model: this.model,
        input_tokens: response.usage.input_tokens,
        output_tokens: response.usage.output_tokens,
        timestamp: new Date().toISOString(),
      })
    );

    return text;
  }

  async run(id: string): Promise<TOutput> {
    const input = await this.acquire(id);
    const prompt = this.prepare(input);
    const response = await this.process(prompt, id);
    const output = this.parse(response);
    return await this.render(id, output);
  }

  async runBatch(ids: string[]): Promise<Map<string, TOutput>> {
    const results = new Map<string, TOutput>();
    for (const id of ids) {
      try {
        const output = await this.run(id);
        results.set(id, output);
      } catch (error) {
        console.error(`[error] ${id}:`, error);
      }
    }
    return results;
  }
}

// ============================================================================
// 4. YOUR PIPELINE IMPLEMENTATION
// ============================================================================

export class YourPipeline extends BasePipeline<YourInput, YourOutput> {
  /**
   * Stage 1: Acquire input data from database
   */
  async acquire(itemId: string): Promise<YourInput> {
    // TODO: Replace with your table/query
    const item = await db.query.yourTable.findFirst({
      where: and(
        eq(yourTable.tenant_id, this.tenantId),
        eq(yourTable.id, itemId)
      ),
    });

    if (!item) {
      throw new Error(`Item ${itemId} not found for tenant ${this.tenantId}`);
    }

    return item;
  }

  /**
   * Stage 2: Prepare prompt for LLM
   */
  prepare(input: YourInput): string {
    // TODO: Customize your prompt
    return `Analyze this content.

TITLE: ${input.title}

CONTENT:
${input.content}

I will parse this programmatically. Respond with valid JSON:
{
  "summary": "Brief summary (1-2 sentences)",
  "category": "type_a" | "type_b" | "type_c",
  "confidence": 0.0 to 1.0
}

Ensure JSON is complete and parseable.`;
  }

  /**
   * Stage 4: Parse LLM response
   */
  parse(response: string): YourOutput {
    const jsonMatch = response.match(/\{[\s\S]*\}/);
    if (!jsonMatch) {
      throw new Error("No JSON found in response");
    }

    try {
      const parsed = JSON.parse(jsonMatch[0]);
      return YourOutputSchema.parse(parsed);
    } catch (error) {
      throw new Error(`Failed to parse response: ${error}`);
    }
  }

  /**
   * Stage 5: Render/save output to database
   */
  async render(itemId: string, output: YourOutput): Promise<YourOutput> {
    // TODO: Replace with your update query
    await db
      .update(yourTable)
      .set({
        summary: output.summary,
        category: output.category,
        processed_at: new Date(),
        updated_at: new Date(),
      })
      .where(
        and(eq(yourTable.tenant_id, this.tenantId), eq(yourTable.id, itemId))
      );

    return output;
  }
}

// ============================================================================
// 5. FACTORY FUNCTION
// ============================================================================

export function createYourPipeline(
  tenantId: string,
  options?: { model?: string }
) {
  return new YourPipeline(tenantId, options);
}

// ============================================================================
// 6. USAGE EXAMPLE
// ============================================================================

/*
// In API route or server function:
import { createYourPipeline } from "@/lib/pipelines/your-pipeline";

export async function processItem(tenantId: string, itemId: string) {
  const pipeline = createYourPipeline(tenantId);
  return await pipeline.run(itemId);
}

// Batch processing:
export async function processItems(tenantId: string, itemIds: string[]) {
  const pipeline = createYourPipeline(tenantId);
  return await pipeline.runBatch(itemIds);
}
*/
